Distribution of detector blocks across MPI processes

anand-avinash opened the pull request:
This pull request is related to the discussion #325.

Here I have added a new method `_make_detector_blocks()` to the `Observation` class. It takes a list of the detectors and divides them into different groups as specified by `det_blocks_attributes` parameter passed to the `Observation` class. The group of detectors are saved in the variable `detector_blocks` of `Observation` class. For instance, with `det_blocks_attributes = ["wafer", "pixel"]`, the detector groups will be made in such a way that each group will have detectors with same wafer and pixel attributes.

Once the detectors are divided into the groups, it sets the number of detector blocks. The number of time blocks are computed accordingly (`n_blocks_time = comm.size // n_blocks_det`). Given the structure of the code base, this scheme was pretty easy to implement than I thought, and it doesn't breaks any other functionality.

On the other side, this implementation can lead to load imbalance in different MPI processes. Consider a simulation with 2 MPI processes and 4 detectors. If the detectors are ended up in two groups, with 3 and 1 elements respectively, one MPI process will have to work with 3 times more data than the other MPI process.

Another caveat is related to the `det_idx` member of `Observation` class. Previously, it corresponded to the global list of detectors passed to the `Observation` class. Now with the introduction of detector blocks, it won't be true anymore. For a simple solution, I have introduced a member variable `detectors_global`, that is the global list of detectors sorted according the detector groups. Now the `det_idx` will correspond to the indices of the detectors in `detectors_global`. But then we end up with three variables `detectors` (the list of detectors used to initialize the `Observation` class), `Observation.detector_blocks`, and `Observation.detectors_global` storing the same data in different manners.

After adding documentation, this pull request should become ready to be merged. But let me know if you have other solutions.

ziotom78 commented on litebird_sim/distribute.py:
```diff
@@ -84,6 +84,45 @@ def distribute_evenly(num_of_elements, num_of_groups):
 
 
+def distribute_detector_blocks(detector_blocks):
+    """Similar to the function :func:`distribute_evenly()`, this function returns the named-tuples of the starting index of the detectors in a group
+    with respect to the global list of detectors and the number of detectors
+    in the group. Unlike the :func:`distribute_evenly()`, this function simply
+    uses the detector groups given in `detector_blocks` attribute.
```
Clearer:

> Similar to the distribute_evenly() function, this function returns a list of named tuples with fields `start_idx` (the starting index of the detector in a group within the global list of detectors) and `num_of_elements` (the number of detectors in the group). Unlike distribute_evenly(), this function uses the detector groups given in the `detector_blocks` attribute.

ziotom78 commented on litebird_sim/observations.py:
```diff
@@ -80,11 +82,19 @@ class Observation:
 
         sampling_rate_hz (float): The sampling frequency, in Hertz.
 
+        det_blocks_attributes (list of strings): The list of detector attributes that
+            will be used to divide detector axis of the tod (and all their attributes).
+            For example, with ``det_blocks_attributes = ["wafer", "pixel"]``, the
+            detectors will be divided into the blocks such that all detectors in a
+            block will have same ``wafer`` and ``pixel`` attribute.
```
Fix the grammar:

> The list of detector attributes that will be used to divide the detector axis of the tod matrix and all its attributes. For example, with `det_blocks_attributes = ["wafer", "pixel"]`, the detectors will be divided into blocks such that all detectors in a block will have the same ``wafer`` and ``pixel`` attributes.

ziotom78 commented on litebird_sim/observations.py:
```diff
@@ -80,11 +82,19 @@ class Observation:
+
         n_blocks_det (int): divide the detector axis of the tod (and all the
-            arrays of detector attributes) in `n_blocks_det` blocks
+            arrays of detector attributes) in `n_blocks_det` blocks. Will be ignored
+            if ``det_blocks_attributes`` is not `None`.
```
“It will be ignored…”

ziotom78 commented on litebird_sim/observations.py:
```diff
@@ -80,11 +82,19 @@ class Observation:
 
         n_blocks_time (int): divide the time axis of the tod in
-            `n_blocks_time` blocks
+            `n_blocks_time` blocks. Will be ignored if ``det_blocks_attributes``
+            is not `None`.
```
“It will be ignored…”

ziotom78 commented on litebird_sim/observations.py:
```diff
@@ -273,10 +300,86 @@ def n_blocks_time(self):
         return self._n_blocks_det
 
+    def _make_detector_blocks(self, detectors, comm):
+        """This function distributes the detectors in groups such that each
+        group has same set of attributes specified by the strings in
+        `self._det_block_attributes`. Once the groups are made, number of
+        detector blocks is set to be the total number of detector groups
+        whereas the number of time blocks is computed using the number of
+        detector blocks and the size of `comm` communicator.
+
+        The blocks of detectors are stored in `self.detector_blocks`. It is a
+        dictionary object with the tuple of `self._det_blocks_attributes` values
+        as dictionary keys and the list of detectors corresponding to the key
+        as the dictionary value. This dictionary is sorted so that that the
+        group with largest number of detectors comes first and the one with
+        the least number of detectors, comes last.
```
Better:

> This function distributes the detectors in groups such that each group has the same set of attributes specified by the strings in `self._det_block_attributes`. Once the groups are made, the number of detector blocks is set to be the total number of detector groups, whereas the number of time blocks is computed using the number of detector blocks and the size of the `comm` communicator.
The detector blocks are stored in `self.detector_blocks`. This dictionary object has the tuple of `self._det_blocks_attributes` values as dictionary keys and the list of detectors corresponding to the key as dictionary values. This dictionary is sorted so that the group with the largest number of detectors comes first and the one with the fewest detectors comes last.

ziotom78 commented on litebird_sim/simulations.py:
```diff
@@ -922,7 +923,12 @@ def create_observations(
         this means that each observation will handle ``10 / 5 = 2``
         detectors. The default is that *all* the detectors be kept
-        together (``n_blocks_det=1``).
+        together (``n_blocks_det=1``). On the other hand, the parameter
+        `det_blocks_attributes` specifies the list of detector attributes
+        to be used to create the groups of detectors. For example, with
+        ``det_blocks_attributes = ["wafer", "pixel"]``, the detectors will
+        be divided into the groups such that all detectors in a group will
+        have same ``wafer`` and ``pixel`` attribute.
```
Better:

> On the other hand, the parameter `det_blocks_attributes` specifies the list of detector attributes to create groups of detectors. For example, with ``det_blocks_attributes = ["wafer", "pixel"]``, the detectors will be divided into groups such that all detectors in a group will have the same ``wafer`` and ``pixel``attributes.

ziotom78 reviewed (commented):
Thanks for having fixed the comments!

anand-avinash commented:
I believe this PR is ready to be merged, but there are a few minor issues that need discussion. The implementation of attribute-based detector blocks works well in the full simulation pipeline when `comm.size` equals `n_blocks_det * n_blocks_time`. However, we encounter errors when `comm.size` is larger, as this leaves some MPI processes unused. The `Observation` classes on the unused processes are still being queried at several places, leading to the errors. When we provide `det_blocks_attributes`, the `n_blocks_time` is computed during runtime using the value of `comm.size` (`n_blocks_time = comm.size // n_blocks_det`). So we cannot avoid this situation unless we do the simulations with a well-balanced set of detectors.

Nonetheless, I tried to resolve some of the error which I am documenting here.

The first error I encountered was:
```
Traceback (most recent call last):
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/BrahMap/test_examples/cross_talk/detector_blocks_full_example.py", line 127, in <module>
    binned_maps = lbs.make_binned_map(nside, sim.observations)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/binner.py", line 325, in make_binned_map
    detector_mask_list = _build_mask_detector_split(detector_split, obs_list)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/common.py", line 515, in _build_mask_detector_split
    detector_mask.append(np.ones(cur_obs.n_detectors, dtype=bool))
                                 ^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/observations.py", line 212, in n_detectors
    return len(self.det_idx)
           ^^^^^^^^^^^^^^^^^
TypeError: object of type 'NoneType' has no len()
```
which I solved simply by adding the following:
https://github.com/litebird/litebird_sim/blob/a941fd0f7fbf72f441c17dd4e2edfe732fb9909b/litebird_sim/observations.py#L209-L212

---

The next error I encountered was
```
Traceback (most recent call last):
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/BrahMap/test_examples/cross_talk/detector_blocks_full_example.py", line 127, in <module>
    binned_maps = lbs.make_binned_map(nside, sim.observations)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/binner.py", line 329, in make_binned_map
    nobs_matrix = _build_nobs_matrix(
                  ^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/binner.py", line 209, in _build_nobs_matrix
    cur_weights = get_map_making_weights(cur_obs, check=True)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/common.py", line 107, in get_map_making_weights
    weights = observations.sampling_rate_hz * observations.net_ukrts**2
                                              ~~~~~~~~~~~~~~~~~~~~~~^^~
TypeError: unsupported operand type(s) for ** or pow(): 'NoneType' and 'int'
```
It appears because `setattr_det_global()` sets the detector attributes to `None` on the processes that are unused. I was able to fix it by setting the numerical attributes to `0` instead of `None`:
https://github.com/litebird/litebird_sim/blob/a941fd0f7fbf72f441c17dd4e2edfe732fb9909b/litebird_sim/observations.py#L677-L682

---
The next error I encountered is
```
Traceback (most recent call last):
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/BrahMap/test_examples/cross_talk/detector_blocks_full_example.py", line 127, in <module>
    binned_maps = lbs.make_binned_map(nside, sim.observations)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/binner.py", line 329, in make_binned_map
    nobs_matrix = _build_nobs_matrix(
                  ^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/binner.py", line 224, in _build_nobs_matrix
    pixidx_all, polang_all = _compute_pixel_indices(
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/common.py", line 222, in _compute_pixel_indices
    del curr_pointings_det
        ^^^^^^^^^^^^^^^^^^
UnboundLocalError: cannot access local variable 'curr_pointings_det' where it is not associated with a value
```
It can be solved by replacing line 222 of `litebird_sim/common.py` by following:
```
        try:
            del curr_pointings_det
        except UnboundLocalError:
            pass
```

---

Next one I encountered while calling the destriper:

```
Traceback (most recent call last):
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/destriper.py", line 1619, in make_destriped_map
    len(params.samples_per_baseline)
TypeError: object of type 'int' has no len()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/BrahMap/test_examples/cross_talk/detector_blocks_full_example.py", line 131, in <module>
    destriped_maps = lbs.make_destriped_map(nside, sim.observations)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/destriper.py", line 1630, in make_destriped_map
    baseline_lengths_list = [
                            ^
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/destriper.py", line 1631, in <listcomp>
    split_items_evenly(
  File "/mnt/Data/Projects/uniroma2/coding/dev_cosmomap2/lbsim_v0.13.0/litebird_sim/mapmaking/destriper.py", line 101, in split_items_evenly
    assert sub_n < n, "sub_n={0} is not smaller than n={1}".format(sub_n, n)
           ^^^^^^^^^
AssertionError: sub_n=100 is not smaller than n=0
```
This one seems little non-trivial to solve.

At the end we may face errors elsewhere as well that I have not tested. And it brings me to the question: should we worry about fixing this issue or should we leave it to the user to supply a well-balanced set of detectors?

ziotom78 commented:
> At the end we may face errors elsewhere as well that I have not tested. And it brings me to the question: should we worry about fixing this issue or should we leave it to the user to supply a well-balanced set of detectors?

The simplest solution would probably be to check in `Simulation.__init__` and crash if the number of MPI processes is too large. I’m not sure that this is the best option, we should ask the opinion of people of the Simulation Production Team.

anand-avinash commented:
As we discussed in the telecon, I have now added a new subcommunicator that excludes the processes from `MPI.COMM_WORLD` where there are no detectors. It includes all the processes for which `rank < n_blocks_det*n_blocks_time`. The new subcommunicator is encapsulated in a class instance (`litebird_sim.mpi.comm_grid`) defined in `mpi.py`, with explicitly defined setter function so that it cannot be overwritten accidently. The new subcommunicator can be accessed with `litebird_sim.mpi.comm_grid.COMM_OBS_GRID`. By default, this subcommnicator is equal to `MPI.COMM_WORLD` unless the observations are defined. Depending on the number of detector blocks and time blocks, the value of the subcommunicator is updated.

For the processes that do not belong to this new sub-communicator, the sub-communicator object points to a NULL communicator (`MPI.COMM_NULL`). When using `litebird_sim.mpi.COMM_OBS_GRID`, one must check if it is not same as the NULL communicator. The general NULL communicator can be accessed with `litebird_sim.mpi.comm_grid.COMM_NULL`.

To resolve the issues I reported previously, I have replaced the global communicator with the subcommunicator at various places. However, there might be some other places where this replacement is necessary, for example `io.py`, `hwp_sys.py`, etc. We can update them if someone encounters an error in future.

@ziotom78 @paganol please let me know if we can proceed with this solution.

ziotom78 commented on litebird_sim/mpi.py:
```diff
@@ -22,10 +22,47 @@ class _SerialMpiCommunicator:
 #: that defines the member variables `rank = 0` and `size = 1`.
 MPI_COMM_WORLD = _SerialMpiCommunicator()
 
+comm_grid = _GridCommClass()
+
```
Being a global variable that is not expected to be reassigned, I would name it `MPI_COMM_GRID`, in analogy with `MPI_COMM_WORLD`, `MPI_ENABLED`, etc.

ziotom78 replied:
Refactoring a variable name is trivial if you use an advanced IDE like PyCharm. If you don't, let me know, and I'll rename it. (It’s just a matter of one mouse click.)

anand-avinash replied:
> Being a global variable that is not expected to be reassigned, I would name it `MPI_COMM_GRID`, in analogy with `MPI_COMM_WORLD`, `MPI_ENABLED`, etc.

Do you mean removing the `_GridCommClass` altogether and instead simply use `MPI_COMM_GRID`?

ziotom78 replied:
> > Being a global variable that is not expected to be reassigned, I would name it `MPI_COMM_GRID`, in analogy with `MPI_COMM_WORLD`, `MPI_ENABLED`, etc.
> 
> Do you mean removing the `_GridCommClass` altogether and instead simply use `MPI_COMM_GRID`?

No, it's just a matter of renaming `comm_grid`, as in this way it will match the naming convention of the other global variables in the `mpi` module.

ziotom78 reviewed (changes requested):
I like it! Nice work!

ziotom78 commented:
My gosh, thank you so much for having spotted this, @anand-avinash ! It was added four years ago but never documented.

> It might be useful to make the sub-communicators for detector and time blocks the members of Observation class. The detector block sub-communicator can be used to pick up all the ranks where the TOD chunks of a given detector reside. Similarly the time block sub-communicator can be used to pick all the ranks where TOD chunks of a given time interval is located.

Absolutely, I agree!
